doc/man3/OSSL_STORE_open.pod

```diff
@@ -143,8 +143,8 @@ on error or when end of data is reached.
 Use OSSL_STORE_error() and OSSL_STORE_eof() to determine the meaning of a
 returned NULL.
 
-OSSL_STORE_eof() returns 1 if the end of data has been reached, otherwise
-0.
+OSSL_STORE_eof() returns 1 if the end of data has been reached, 0 if not,
```

Comment:
It would be really bad to change the API this way. It was not expected to return <0 on error in 1.1.1.

Reply:
This is not an API change, but a correction of the documentation!
Note that `BIO_eof()` and thus `OSSL_STORE_eof()` CAN return -1 or -2 on error.
Or shall solve the discrepancy the other way round, making sure that <0 is mapped to 0?

Reply:
OSSL_STORE_eof() should never return < 0.  I hadn't thought of what BIO_eof() can return...

Here's my suggestion:

``` diff
diff --git a/providers/implementations/storemgmt/file_store.c b/providers/implementations/storemgmt/file_store.c
index 37f2fcee67..c300698b8c 100644
--- a/providers/implementations/storemgmt/file_store.c
+++ b/providers/implementations/storemgmt/file_store.c
@@ -705,12 +705,23 @@ static int file_eof(void *loaderctx)
     case IS_DIR:
         return ctx->_.dir.end_reached;
     case IS_FILE:
+        int ret;
+
         /*
-         * BIO_pending() checks any filter BIO.
-         * BIO_eof() checks the source BIO.
+         * BIO_pending() checks any filter BIO.  If anything is pending, or
+         * there was an error, this can't be EOF.
          */
-        return !BIO_pending(ctx->_.file.file)
-            && BIO_eof(ctx->_.file.file);
+        ret = BIO_pending(ctx->_.file.file);
+        if (ret != 0)
+            return 0;
+
+        /* BIO_eof() checks the source BIO. */
+        ret = BIO_eof(ctx->_.file.file);
+        if (ret <= 0)
+            return 0;
+
+        /* We have reached the end of input */
+        return 1;
     }
 
     /* ctx->type has an unexpected value */
```

Reply:
This can happen only if the CTRL is unimplemented, isn't it so?

Reply:
BIO_eof is not documented to return -1 or -2 either. So it should be fixed instead.

Reply:
IMO it returns -1 or -2 only on _impossible_ conditions.

Reply:
> So it [BIO_eof()] should be fixed instead.

I dunno about that, the BIO_eof() behavior is much older (you can trace if *far* back), and coded directly.  That the file scheme implementation of OSSL_STORE_eof() happened to propagate it was accidental.

So, while you might want to fix BIO_eof(), the patch I proposed is still not *wrong*.

Reply:
How about adapting in `bio.h.in`:
```
# define BIO_eof(b)              (int)BIO_ctrl(b,BIO_CTRL_EOF,0,NULL)
```
to
```
# define BIO_eof(b)              (BIO_ctrl(b,BIO_CTRL_EOF,0,NULL) == 1)
```
This way `file_eof()` and `OSSL_STORE_eof()` get fixed as well.

Reply:
I've just executed my proposal.